src/app/modules/spcp/sp.oidc.client.ts

```diff
@@ -0,0 +1,583 @@
+ * Handles discovery and retrieval of NDI's public jwks
+ * Exported for testing
+ */
+export class SpOidcClientCache extends NodeCache {
+  #spOidcNdiDiscoveryEndpoint: string
+  #spOidcNdiJwksEndpoint: string
+  #spOidcRpClientId: string
+  #spOidcRpRedirectUrl: string
+  #spOidcRpSecretJwks: SecretJwks
+
+  /**
+   * Constructor for cache
+   * On instantiation, trigger a refresh to populate the cache
+   */
+  constructor({
+    options,
+    spOidcNdiDiscoveryEndpoint,
+    spOidcNdiJwksEndpoint,
+    spOidcRpClientId,
+    spOidcRpRedirectUrl,
+    spOidcRpSecretJwks,
+  }: SpOidcClientCacheConstructorParams) {
+    super(options)
+
+    this.#spOidcNdiDiscoveryEndpoint = spOidcNdiDiscoveryEndpoint
+    this.#spOidcNdiJwksEndpoint = spOidcNdiJwksEndpoint
+    this.#spOidcRpClientId = spOidcRpClientId
+    this.#spOidcRpRedirectUrl = spOidcRpRedirectUrl
+    this.#spOidcRpSecretJwks = spOidcRpSecretJwks
+
+    // On expiry, refresh cache. If fail to refresh, log but do not throw error.
+    this.on('expired', () =>
+      this.refresh().catch((err) =>
+        console.warn(
+          `Attempted but failed to refresh sp oidc cache on expiry. Reason: ${err}`,
+        ),
+      ),
+    )
+
+    // Trigger refresh on instantiation to populate cache. If fail to refresh, log but do not throw error.
+    void this.refresh().catch((err) =>
+      console.warn(
+        `Attempted but failed to refresh sp oidc on instantiation. Reason: ${err}`,
+      ),
+    )
+  }
+
+  /**
+   * Method to retrieve NDI's public keys from cache
+   * @async
+   * @returns NDI's public keys
+   */
+  async getNdiPublicKeys(): Promise<CryptoKeySet> {
+    const ndiPublicKeys = this.get<CryptoKeySet>('ndiPublicKeys')
+    if (!ndiPublicKeys) {
+      const { ndiPublicKeys } = await this.refresh()
+      return ndiPublicKeys
+    }
+    return ndiPublicKeys
+  }
+
+  /**
+   * Method to retrieve base client from cache
+   * Triggers a refresh
+   * @returns Base client
+   */
+  async getBaseClient(): Promise<BaseClient> {
+    const baseClient = this.get<BaseClient>('baseClient')
+    if (!baseClient) {
+      const { baseClient } = await this.refresh()
+      return baseClient
+    }
+    return baseClient
+  }
+
+  /**
+   * Method to trigger a refresh and fetch NDI's public keys and
+   * discover the well known endpoint to construct the base client
+   * Stores NDI's public key and base client in cache with a TTL of 60 mins
+   * Sets `expiry` key in cache with TTL of 59 mins to trigger refresh ahead
+   * @returns object {ndiPublicKeys, baseClient}
+   */
+  async refresh(): Promise<Refresh> {
+    const [ndiPublicKeys, baseClient] = await Promise.all([
+      this.retrievePublicKeysFromNdi(),
+      this.retrieveBaseClientFromNdi(),
+    ])
+
+    this.set('ndiPublicKeys', ndiPublicKeys, 3600) // TTL of 60 minutes
+    this.set('baseClient', baseClient, 3600) // TTL of 60 minutes
```

Comment:
To explain the thinking here - if the refresh ahead fails (max 3 attempts based on specs), then that means we will keep the keys indefinitely until our interaction with NDI fails subsequently (which indicates keys no longer valid). 

Hence, the `ndiPublicKeys` and `baseClient` keys have a 1h TTL so that at that point, another attempt is made to refresh the keys. If the keys are still not available, then we return an error (but try to refetch the keys on every request, which re-sets the TTL).

An alternative I can think of is to not set expiry on the 2 keys, but keep on attempting to refresh indefinitely from the 55min mark until success, with exponential backoff for interval between refresh. Thoughts?

Reply:
Thanks for that ShuLi! The alternative you propose sounds good to me 👍 

To clarify a little further, I had 2 considerations with my comment:
1. As long as we have managed to fetch data, let's optimistically keep it around and reuse it, under the assumption that NDI data just doesn't change all that often anyway.
2. Let's be really careful with parallel refreshes

For 2. With the current code, if the fetch-ahead fails, and the keys then expire, parallel refreshes can be triggered by inbound calls to the app, which is not ideal.

Keeping the keys alive ensures that can't happen (except at startup), and indeed we should have an independent process running to continuously attempt to (cheaply) refresh the keys.

If the code implements an additional guarantees that there can only be a single refresh in-flight at a time. That's even better.